src/mame/sinclair/sprinter.cpp

```diff
@@ -0,0 +1,1183 @@
+			sprinter_mem_w(offset + i, m_accel_buffer[i]);
+	}
+	else if (m_accel_state == COPY_VERT)
+	{
+		LOGACCEL("Accel wCOPY_VERT: %02x (%x)\n", offset, m_port_y);
+		for (auto i = 0; i < m_accel_buffer_size; i++)
+		{
+			sprinter_mem_w(offset, m_accel_buffer[i]);
+			m_port_y++;
+		}
+	}
+	else
+		adjust_icount = false;
+
+	if (!adjust_icount)
+		sprinter_mem_w(offset, data);
+	else
+	{
+		int icount = 6 * m_accel_buffer_size * m_maincpu->clock() / X_SP; // 6 42Mhz clocks each
+		m_maincpu->adjust_icount(-icount);
+	}
+}
+
+void sprinter_state::accel_control_r(u8 data)
+{
+	accel_state state_candidate = static_cast<accel_state>(data);
+	switch(state_candidate)
+	{
+	case OFF_HALT:
+	case OFF:
+		m_accel_state = OFF;
+		LOGACCEL("Accel: OFF\n");
+		break;
+
+	case SET_BUFFER:
+	case FILL:
+	case FILL_VERT:
+	case COPY:
+	case COPY_VERT:
+		m_accel_state = state_candidate;
+		m_accel_mode = MODE_NOP;
+		break;
+
+	case MODE_AND:
+	case MODE_XOR:
+	case MODE_OR:
+		m_accel_mode = state_candidate;
+		break;
+
+	default:
+		m_accel_mode = MODE_NOP;
+		break;
+	}
+}
+
+u8 sprinter_state::vram_r(offs_t offset)
+{
+	return m_ram->pointer()[(0x50 << 14) + m_port_y * 1024 + (offset & 0x3ff)];
+}
+
+template <u8 Bank> void sprinter_state::vram_w(offs_t offset, u8 data)
+{
+	u8 page = m_pages[Bank] & 0xff;
+	u16 vaddr = (offset & 0x03ff);
+	u8* line = m_vram + (m_port_y * 1024);
+
+	if (BIT(~page, 2))
+		m_ram->pointer()[(0x50 << 14) + m_port_y * 1024 + (offset & 0x3ff)] = data;
+	if (!(BIT(page, 3) && (data == 0xff)))
+	{
+		m_screen->update_now();
+		line[vaddr] = data;
+	}
+
+	if (vaddr >= 0x3e0)
+	{
+		u8 pal_num = BIT(vaddr, 2, 3);
+		u16 p_red = 0x3e0 + pal_num * 4;
+		m_palette->set_pen_color(m_port_y + pal_num * 256, rgb_t(line[p_red], line[p_red + 1], line[p_red + 2]));
+	}
+}
+
+template <u8 Bank> u8 sprinter_state::ram_r(offs_t offset)
+{
+	return reinterpret_cast<u8 *>(m_bank_ram[Bank]->base())[offset];
+}
+
+template <u8 Bank> void sprinter_state::ram_w(offs_t offset, u8 data)
+{
+	if (~m_all_mode & 1)
+	{
+		bool sp_scr = BIT(Bank, 0) && (BIT(~Bank, 1) || ((m_pg3 & 0x3d) == 0x35)); // B"1101X1"
+		bool zx_screen = sp_scr && !(BIT(offset, 13) && !BIT(m_port_y, 7)) && !BIT(m_port_y, 6) && !BIT(m_port_y, 0);
+		if (zx_screen)
+		{
+			u8 sp_sa = BIT(Bank, 1) ? BIT(m_pg3, 1) : 0;
+			u32 addr = ((offset & 0xff) << 10)
+				| (((m_port_y & 0x1f) ^ sp_sa ^ BIT(offset, 13)) << 5)
+				| (BIT(offset, 8, 5));
+
+			m_screen->update_now();
+			m_vram[addr] = data;
+		}
+	}
+
+	reinterpret_cast<u8 *>(m_bank_ram[Bank]->base())[offset] = data;
+}
+
+u8 sprinter_state::sprinter_mem_r(u16 offset)
+{
+	accel_state state = m_accel_state;
+	m_accel_state = DISABLED;
+	u8 data = m_program.read_byte(offset);
+	m_accel_state = state;
+
+	return data;
+}
+
+void sprinter_state::sprinter_mem_w(u16 offset, u8 data)
+{
+	accel_state state = m_accel_state;
+	m_accel_state = DISABLED;
+	m_program.write_byte(offset, data);
+	m_accel_state = state;
+}
```

Comment:
This seems like a pretty expensive way of doing this – accesses through these ranges will end up going through the address space dispatch to get here, then it will have to go through another layer of address space dispatch as well as this trampoline.  Is there really not a better way of doing it?